src/objects/Cell.h

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file Cell.h
+ * @brief A single Cell in a CellContainer, for use with the linked cells method.
+ * @date 2024-12-02
+ *
+ * @copyright Copyright (c) 2024
+ *
+ */
+#pragma once
+
+#include "Particle.h"
+#include "ParticleContainer.h"
+#include "utils/CellUtils.h"
+#include <algorithm>
+#include <array>
+#include <bitset>
+#include <forward_list>
+#include <sstream>
+#include <string>
+#include <vector>
+#define VEC_CONTAINS(vec, el) ((std::find((vec).begin(), (vec).end(), (el)) != (vec).end()))
+
+/// @brief Class for storing data of a single Cell in a domain split into cells.
+class Cell {
+    /// @brief Typedef for the container type used to store pointers to a Cell's Particle objects.
+    using ContainerType = std::forward_list<Particle *>;
+
+  private:
+    /// @brief A forward list of pointers to Particle objects contained within the current Cell.
+    ContainerType m_particles{};
+    /// @brief The size of the Cell in each dimension.
+    std::array<double, 3> m_size;
+    /// @brief The lower-left coordinates of the Cell.
+    std::array<double, 3> m_position;
+    /// @brief If this is a halo cell, the locations of this cell (North, South, West, East, Above, Below) are stored
+    /// here.
+    std::vector<HaloLocation> m_haloLocation;
+    /// @brief The type of this Cell. May be INNER, BORDER or HALO.
+    CellType m_type;
+    /// @brief The index of this Cell in the overarching CellContainer.
+    int m_index;
+
+  public:
+    /**
+     * @brief Constructs a new Cell object.
+     *
+     * Typically called from a CellContainer.
+     *
+     * @param size The size of the Cell in each dimension.
+     * @param position The lower-left coordinates of the Cell.
+     * @param type The type of this Cell.
+     * @param index The index of this Cell in the CellContainer.
+     * @param haloLocation The cardinal direction(s) of this cell if this is a halo Cell.
+     */
+    Cell(const std::array<double, 3> &size, const std::array<double, 3> &position, CellType type, int index,
+         const std::vector<HaloLocation> &haloLocation);
+
+    /**
+     * @brief Gets an iterator to the beginning of the Particle* forward list.
+     *
+     * @return An iterator to the beginning of the Cell's forward list.
+     */
+    ContainerType::iterator begin();
+
+    /**
+     * @brief Gets an iterator to one past the end of the Particle* forward list.
+     *
+     * @return An iterator to one past the end of the Cell's forward list.
+     */
+    ContainerType::iterator end();
+
+    /**
+     * @brief Adds a Particle pointer to the front of the forward list.
+     *
+     * @param particle The Particle pointer to be added.
+     */
+    void addParticle(Particle *particle);

```

Comment:
Yeah, that's an old habit from years of C programming. We changed the raw pointers to `std::reference_wrapper` instances now. 

It was a bit difficult trying to figure out which smart pointer was the "correct" choice; using `std::unique_ptr` would be problematic for moving particles from one cell to another (i.e. deleting the pointers from one cell's list and adding them to another cell's list), since we'd have to somehow ensure that the pointers don't get deallocated once the function returns (as in... we'd have to rewrite the methods completely), while using `std::shared_ptr` would result in needless overhead through reference counting and wouldn't really make that much sense semantically, since only a cell is responsible for keeping a pointer to a particle (also, I think it's generally advised _not_ to use shared pointers unless absolutely necessary).

Reply:
I can imagine this being implemented with just a reference, `addParticle(Particle &particle)`. But anyway, this is not a major concern, so don't worry too much about it.